patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+    :param dsn: Optional DSN string for the local instance to get GUC values from.
+    """
+    no_value_msg = '#FIXME'
+    pg_version = None
+    local_ip = get_ip()
+
+    config = Config('', None).local_configuration  # Get values from env
+    dynamic_config = Config.get_default_config()
+    dynamic_config['postgresql']['parameters'] = dict(dynamic_config['postgresql']['parameters'])
+    config.setdefault('bootstrap', {})['dcs'] = dynamic_config
+    config.setdefault('postgresql', {})
+
+    template_config: Dict[str, Any] = {
+        'scope': no_value_msg,
+        'name': socket.gethostname(),
+        'postgresql': {
+            'data_dir': no_value_msg,
+            'connect_address': no_value_msg + ':5432',
+            'listen': no_value_msg + ':5432',
+            'bin_dir': '',
+            'authentication': {
+                'superuser': {
+                    'username': 'postgres',
+                    'password': no_value_msg
+                },
+                'replication': {
+                    'username': 'replicator',
+                    'password': no_value_msg
+                }
+            }
+        },
+        'restapi': {
+            'connect_address': local_ip + ':8008',
+            'listen': local_ip + ':8008'
+        }
+    }
+
+    if not sample:
+        enrich_config_from_running_instance(config, no_value_msg, dsn)
+        config['postgresql']['bin_dir'] = get_bin_dir_from_running_instance(config['postgresql']['data_dir'])
+
+    try:
+        pg_version = get_int_major_version(config)
+    except PatroniException as e:
+        sys.exit(str(e))
+
+    patch_config(template_config, config)
+    config = template_config
+
+    # generate sample config
+    if sample:
+        auth_method = 'scram-sha-256' if pg_version and pg_version >= 100000 else 'md5'
+        config['postgresql']['parameters'] = {'password_encryption': auth_method}
+        config['postgresql']['pg_hba'] = [
+            f'host all all all {auth_method}',
+            f'host replication {config["postgresql"]["authentication"]["replication"]["username"]} all {auth_method}'
+        ]
+
+        # add version-specific configuration
+        wal_keep_param = 'wal_keep_segments' if pg_version < 130000 else 'wal_keep_size'
+        config['bootstrap']['dcs']['postgresql']['parameters'][wal_keep_param] =\
+            ConfigHandler.CMDLINE_OPTIONS[wal_keep_param][0]
+
+        config['bootstrap']['dcs']['postgresql']['use_pg_rewind'] = True
+        if pg_version >= 110000:
+            config['postgresql']['authentication'].setdefault(
+                'rewind', {'username': 'rewind_user'}).setdefault(
+                    'password', no_value_msg)
+
+    # redundant values from the default config
+    del config['bootstrap']['dcs']['standby_cluster']
+
+    if file:
+        dir_path = os.path.dirname(file)
+        if dir_path and not os.path.isdir(dir_path):
+            os.makedirs(dir_path)
+        with open(file, 'w') as fd:
```

Comment:
`open` should have encoding set.

Reply:
I would consider having os modifying code separate from configuration generation and construction code.

Reply:
> open should have encoding set.

had a deep dive into this, thanks:)
https://docs.python.org/3/library/os.html#utf8-mode
still not sure we need to explicitly specify it now, but definitely won't hurt. 

the more serious problem, I guess, is with `safe_dump`, we definitely need to use `allow_unicode=True` there